File: services/kmService.ts
import { SampleData, ModelCoefficients, ModelType } from '../types';
import { pseudoInverse, multiply } from '../utils/matrix';
import { WAVELENGTHS, Rg_DEFAULT } from '../constants';
import { trainNeuralNetwork, predictNeuralNetwork } from '../utils/neuralNet';

// --- Fluorescence Detection ---

/**
 * Detect if a sample contains fluorescent pigments
 * Fluorescent samples have R > 1 at some wavelengths
 */
export const hasFluorescence = (spectrum: number[]): boolean => {
  return spectrum.some(r => r > 1.0);
};

/**
 * Extract fluorescence component from spectrum
 * For fluorescent samples: F(λ) = max(0, R(λ) - 1)
 * This is the excess reflectance due to fluorescence emission
 */
export const extractFluorescence = (spectrum: number[]): {
  reflectance: number[],
  fluorescence: number[]
} => {
  const reflectance = spectrum.map(r => Math.min(r, 1.0));
  const fluorescence = spectrum.map(r => Math.max(0, r - 1.0));
  return { reflectance, fluorescence };
};

// --- Single Layer Logic ---

export const reflectanceToKS = (R: number): number => {
  // For fluorescent materials, clamp to valid range
  const rSafe = Math.max(0.001, Math.min(0.999, R));
  return Math.pow(1 - rSafe, 2) / (2 * rSafe);
};

export const ksToReflectance = (KS: number): number => {
  const term = Math.sqrt(Math.pow(KS, 2) + 2 * KS);
  return 1 + KS - term;
};

// --- Two Layer Logic (From Python Script) ---

const twoLayerForward = (K_mix: number, S_mix: number, Rg: number, X: number): number => {
  const S = Math.max(1e-6, S_mix);
  const K = Math.max(0, K_mix);

  const a = 1 + K / S;
  const b = Math.sqrt(Math.pow(a, 2) - 1);
  
  // Limit exp power to avoid overflow/underflow
  const bSX = Math.max(-50, Math.min(50, b * S * X));
  
  const exp_bSX = Math.exp(bSX);
  const exp_neg = Math.exp(-bSX);

  const num = (1 - Rg * (a - b)) * exp_bSX * (a + b) - (1 - Rg * (a + b)) * exp_neg * (a - b);
  const den = (a + b - Rg * (a - b)) * exp_bSX - (a - b - Rg * (a + b)) * exp_neg;

  if (den === 0) return 0;
  const R = num / den;
  return Math.max(0, Math.min(1, R));
};

// --- Training ---

export const trainModel = (samples: SampleData[], type: ModelType): ModelCoefficients => {
  if (samples.length === 0) throw new Error("No samples to train on.");

  // Filter samples to only those with concentration data
  const samplesWithConc = samples.filter(s => {
    const concs = Object.values(s.concentrations);
    return concs.length > 0 && concs.some(c => c > 0);
  });

  if (samplesWithConc.length === 0) {
    throw new Error("No samples with concentration data found. Cannot train model.");
  }

  console.log(`Filtered to ${samplesWithConc.length} samples with concentration data (from ${samples.length} total)`);

  // Detect fluorescent samples
  const fluorescentSamples = samplesWithConc.filter(s => hasFluorescence(s.spectrum));
  console.log(`Found ${fluorescentSamples.length} fluorescent samples (R > 1.0)`);

  // Separate spectra into reflectance and fluorescence components
  const processedSamples = samplesWithConc.map(s => {
    const { reflectance, fluorescence } = extractFluorescence(s.spectrum);
    return {
      ...s,
      reflectance,
      fluorescence,
      isFluorescent: hasFluorescence(s.spectrum)
    };
  });

  const reagentSet = new Set<string>();
  samplesWithConc.forEach(s => {
    Object.keys(s.concentrations).forEach(k => {
      // Only add reagents with non-zero concentration
      if ((s.concentrations[k] || 0) > 0) {
        reagentSet.add(k);
      }
    });
  });
  const reagents = Array.from(reagentSet).sort();

  console.log(`Training with ${samplesWithConc.length} samples and ${reagents.length} reagents:`, reagents.join(', '));

  // Concentration Matrix [Samples x Reagents]
  const C = samplesWithConc.map(s => reagents.map(r => (s.concentrations[r] || 0) / 100.0));

  if (type === 'single') {
    // Analytical Solution for K/S using reflectance component only
    const KS_measured = processedSamples.map(s => s.reflectance.map(r => reflectanceToKS(r)));

    try {
      const C_pinv = pseudoInverse(C);
      const Alpha = multiply(C_pinv, KS_measured);

      const alphaMap: Record<string, number[]> = {};
      reagents.forEach((r, idx) => {
        alphaMap[r] = Alpha[idx];
      });

      // Train fluorescence model if fluorescent samples exist
      let fluorMap: Record<string, number[]> | undefined = undefined;
      if (fluorescentSamples.length > 0) {
        console.log("Training fluorescence emission model...");
        const F_measured = processedSamples.map(s => s.fluorescence);
        const Fluor = multiply(C_pinv, F_measured);

        fluorMap = {};
        reagents.forEach((r, idx) => {
          fluorMap![r] = Fluor[idx];
        });
        console.log("Fluorescence model trained successfully");
      }

      console.log("Single-layer model trained successfully");
      return {
        type: 'single',
        alpha: alphaMap,
        fluorescence: fluorMap,
        wavelengths: WAVELENGTHS
      };
    } catch (err) {
      console.error("Failed to compute pseudo-inverse:", err);
      throw new Error("Failed to train single-layer model. Try using fewer samples or check data quality.");
    }
  } else if (type === 'two-layer') {
    // Two-Layer Optimization (Numerical Gradient Descent)
    return trainTwoLayerModel(processedSamples, reagents, C, fluorescentSamples.length > 0);
  } else {
    // Neural Network (Black-box model)
    return trainNeuralNetModel(samplesWithConc, reagents, C);
  }
};

const trainTwoLayerModel = (
  samples: any[], // processed samples with reflectance/fluorescence
  reagents: string[],
  C: number[][],
  hasFluorescent: boolean
): ModelCoefficients => {
  const numWavelengths = WAVELENGTHS.length;
  const numReagents = reagents.length;
  
  const K_map: Record<string, number[]> = {};
  const S_map: Record<string, number[]> = {};
  const F_map: Record<string, number[]> = {};

  reagents.forEach(r => {
    K_map[r] = new Array(numWavelengths).fill(0);
    S_map[r] = new Array(numWavelengths).fill(0);
    if (hasFluorescent) {
      F_map[r] = new Array(numWavelengths).fill(0);
    }
  });

  // Perform optimization for each wavelength independently
  for (let w = 0; w < numWavelengths; w++) {
    // Target Reflectances for this wavelength (use reflectance component only)
    const R_targets = samples.map(s => s.reflectance[w]);
    
    // Thicknesses
    const X_vals = samples.map(s => s.thickness);

    // Initial Guess: Better starting values based on typical K-M values
    // For most pigments: K ~ 0.5-5.0, S ~ 5.0-50.0
    // Params vector: [K_1...K_n, S_1...S_n]
    let params = [...new Array(numReagents).fill(1.0), ...new Array(numReagents).fill(10.0)];

    // Adaptive Gradient Descent with momentum
    const initialLR = 0.001;
    const iterations = 200;
    const momentum = 0.9;
    let velocity = new Array(2 * numReagents).fill(0); 

    for (let iter = 0; iter < iterations; iter++) {
      const grads = new Array(2 * numReagents).fill(0);
      
      // Compute Gradients
      for (let sIdx = 0; sIdx < samples.length; sIdx++) {
        const concs = C[sIdx]; // Vector of concentrations
        const X = X_vals[sIdx];
        const R_tgt = R_targets[sIdx];
        
        // Current K_mix, S_mix
        let K_mix = 0, S_mix = 0;
        for (let r = 0; r < numReagents; r++) {
          K_mix += concs[r] * params[r];
          S_mix += concs[r] * params[r + numReagents];
        }

        const R_pred = twoLayerForward(K_mix, S_mix, Rg_DEFAULT, X);
        const error = R_pred - R_tgt;

        // Finite difference approximation for gradient 
        // (Simplifies implementation vs analytical derivation)
        const epsilon = 1e-4;
        
        for (let r = 0; r < numReagents; r++) {
           // dL/dK_r
           // Approximation: dR/dK * conc
           const R_k_plus = twoLayerForward(K_mix + epsilon * concs[r], S_mix, Rg_DEFAULT, X);
           const dR_dK = (R_k_plus - R_pred) / epsilon;
           grads[r] += 2 * error * dR_dK;

           // dL/dS_r
           const R_s_plus = twoLayerForward(K_mix, S_mix + epsilon * concs[r], Rg_DEFAULT, X);
           const dR_dS = (R_s_plus - R_pred) / epsilon;
           grads[r + numReagents] += 2 * error * dR_dS;
        }
      }

      // Normalize gradients to prevent explosion
      const gradNorm = Math.sqrt(grads.reduce((sum, g) => sum + g * g, 0));
      const normalizedGrads = gradNorm > 1 ? grads.map(g => g / gradNorm) : grads;

      // Update Params with momentum
      const learningRate = initialLR / (1 + iter / 100); // Decay learning rate
      for (let i = 0; i < params.length; i++) {
        velocity[i] = momentum * velocity[i] - learningRate * normalizedGrads[i];
        params[i] += velocity[i];

        // Enforce bounds (positive coefficients with minimum values)
        // K should be >= 0, S should be > 0
        if (i < numReagents) {
          params[i] = Math.max(0.01, params[i]); // K values
        } else {
          params[i] = Math.max(0.1, params[i]);  // S values
        }
      }

      // Early stopping if gradients are very small
      if (gradNorm < 1e-6 && iter > 50) {
        break;
      }
    }

    // Log optimization progress for this wavelength
    if (w % 10 === 0) {
      const avgK = params.slice(0, numReagents).reduce((a, b) => a + b, 0) / numReagents;
      const avgS = params.slice(numReagents).reduce((a, b) => a + b, 0) / numReagents;
      console.log(`λ=${WAVELENGTHS[w]}nm: avg K=${avgK.toFixed(3)}, avg S=${avgS.toFixed(3)}`);
    }

    // Store optimized values
    reagents.forEach((r, rIdx) => {
      K_map[r][w] = params[rIdx];
      S_map[r][w] = params[rIdx + numReagents];
    });

    // Train fluorescence separately if needed
    if (hasFluorescent) {
      const F_targets = samples.map(s => s.fluorescence[w]);
      // Simple linear regression for fluorescence emission
      // F = sum(c_i * F_i) where F_i is fluorescence coefficient per reagent
      const C_pinv = pseudoInverse(C);
      const F_coeffs = multiply(C_pinv, F_targets.map(f => [f])).map(row => row[0]);
      reagents.forEach((r, rIdx) => {
        F_map[r][w] = Math.max(0, F_coeffs[rIdx]); // Fluorescence must be non-negative
      });
    }
  }

  console.log("Two-layer model trained successfully");

  return {
    type: 'two-layer',
    K: K_map,
    S: S_map,
    fluorescence: hasFluorescent ? F_map : undefined,
    wavelengths: WAVELENGTHS
  };
};

const trainNeuralNetModel = (
  samples: SampleData[],
  reagents: string[],
  C: number[][]
): ModelCoefficients => {
  console.log("Training neural network (black-box model)...");

  // Prepare training data
  // Input: [concentration1, concentration2, ..., thickness]
  // Output: [R400, R410, ..., R700]

  const X: number[][] = samples.map((s, idx) => {
    return [...C[idx], s.thickness];
  });

  const Y: number[][] = samples.map(s => s.spectrum);

  // Train neural network with improved hyperparameters
  const {
    weights1,
    bias1,
    weights2,
    bias2,
    inputMean,
    inputStd,
    outputMean,
    outputStd
  } = trainNeuralNetwork(
    X,
    Y,
    128,  // Hidden layer size (increased for better capacity)
    0.005, // Learning rate (reduced for stability)
    2000, // Epochs (increased for convergence)
    Math.min(8, Math.max(4, Math.floor(samples.length / 5))), // Batch size
    0.005 // L2 regularization (prevents overfitting)
  );

  return {
    type: 'neural-net',
    neuralNet: {
      reagents,
      weights1,
      bias1,
      weights2,
      bias2,
      inputMean,
      inputStd,
      outputMean,
      outputStd
    },
    wavelengths: WAVELENGTHS
  };
};

export const predictReflectance = (
  concentrations: Record<string, number>,
  model: ModelCoefficients,
  thickness: number = 4.0
): number[] => {
  const { type, wavelengths } = model;

  if (type === 'neural-net' && model.neuralNet) {
    // Neural network prediction
    const { reagents, weights1, bias1, weights2, bias2, inputMean, inputStd, outputMean, outputStd } = model.neuralNet;

    // Prepare input: [conc1, conc2, ..., thickness]
    const input = reagents.map(r => (concentrations[r] || 0) / 100.0);
    input.push(thickness);

    return predictNeuralNetwork(
      input,
      weights1,
      bias1,
      weights2,
      bias2,
      inputMean,
      inputStd,
      outputMean,
      outputStd
    );
  }

  // K-M based predictions
  const predictedR = new Array(wavelengths.length).fill(0);

  for (let i = 0; i < wavelengths.length; i++) {
    let baseReflectance = 0;

    if (type === 'single') {
      // Single Layer K-M
      let ksSum = 0;
      Object.entries(concentrations).forEach(([reagent, percent]) => {
        const unitKS = model.alpha?.[reagent]?.[i] || 0;
        ksSum += (percent / 100.0) * unitKS;
      });
      ksSum = Math.max(0.0001, ksSum);
      baseReflectance = ksToReflectance(ksSum);
    } else {
      // Two-Layer K-M
      let kSum = 0;
      let sSum = 0;
      Object.entries(concentrations).forEach(([reagent, percent]) => {
        const unitK = model.K?.[reagent]?.[i] || 0;
        const unitS = model.S?.[reagent]?.[i] || 0;
        kSum += (percent / 100.0) * unitK;
        sSum += (percent / 100.0) * unitS;
      });
      baseReflectance = twoLayerForward(kSum, sSum, Rg_DEFAULT, thickness);
    }

    // Add fluorescence contribution if model has fluorescence coefficients
    let fluorescence = 0;
    if (model.fluorescence) {
      Object.entries(concentrations).forEach(([reagent, percent]) => {
        const unitF = model.fluorescence?.[reagent]?.[i] || 0;
        fluorescence += (percent / 100.0) * unitF;
      });
    }

    // Total reflectance = base reflectance + fluorescence emission
    predictedR[i] = baseReflectance + fluorescence;
  }

  return predictedR;
};

File: python_version/services/km_service.py
"""
Kubelka-Munk model training and prediction service
Supports single-layer, two-layer, and neural network models
"""
import numpy as np
from typing import List, Dict, Tuple
import sys
import os

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(__file__)))

from types_constants import SampleData, ModelCoefficients, WAVELENGTHS, Rg_DEFAULT
from utils.matrix_ops import pseudo_inverse, solve_least_squares
from models.neural_network import train_neural_network, predict_neural_network


# --- Fluorescence Detection ---

def has_fluorescence(spectrum: np.ndarray) -> bool:
    """
    Detect if a sample contains fluorescent pigments
    Fluorescent samples have R > 1 at some wavelengths

    Args:
        spectrum: Reflectance spectrum

    Returns:
        True if fluorescence detected
    """
    return np.any(spectrum > 1.0)


def extract_fluorescence(spectrum: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
    """
    Extract fluorescence component from spectrum
    For fluorescent samples: F(λ) = max(0, R(λ) - 1)

    Args:
        spectrum: Original reflectance spectrum

    Returns:
        Tuple of (reflectance, fluorescence) components
    """
    reflectance = np.minimum(spectrum, 1.0)
    fluorescence = np.maximum(0, spectrum - 1.0)
    return reflectance, fluorescence


# --- Single Layer K-M Functions ---

def reflectance_to_ks(R: np.ndarray) -> np.ndarray:
    """
    Convert reflectance to K/S ratio using Kubelka-Munk equation

    Args:
        R: Reflectance values

    Returns:
        K/S ratios
    """
    # Clamp to valid range for fluorescent materials
    R_safe = np.clip(R, 0.001, 0.999)
    return (1 - R_safe) ** 2 / (2 * R_safe)


def ks_to_reflectance(KS: np.ndarray) -> np.ndarray:
    """
    Convert K/S ratio to reflectance using inverse K-M equation

    Args:
        KS: K/S ratio values

    Returns:
        Reflectance values
    """
    term = np.sqrt(KS ** 2 + 2 * KS)
    return 1 + KS - term


# --- Two Layer K-M Functions ---

def two_layer_forward(K_mix: float, S_mix: float, Rg: float, X: float) -> float:
    """
    Two-layer Kubelka-Munk forward model

    Args:
        K_mix: Absorption coefficient
        S_mix: Scattering coefficient
        Rg: Background reflectance
        X: Film thickness

    Returns:
        Predicted reflectance
    """
    S = max(1e-6, S_mix)
    K = max(0, K_mix)

    a = 1 + K / S
    b = np.sqrt(a ** 2 - 1)

    # Limit exp power to avoid overflow/underflow
    bSX = np.clip(b * S * X, -50, 50)

    exp_bSX = np.exp(bSX)
    exp_neg = np.exp(-bSX)

    num = (1 - Rg * (a - b)) * exp_bSX * (a + b) - (1 - Rg * (a + b)) * exp_neg * (a - b)
    den = (a + b - Rg * (a - b)) * exp_bSX - (a - b - Rg * (a + b)) * exp_neg

    if den == 0:
        return 0

    R = num / den
    return np.clip(R, 0, 1)


# --- Training Functions ---

def train_model(samples: List[SampleData], model_type: str) -> ModelCoefficients:
    """
    Train spectral prediction model

    Args:
        samples: List of training samples
        model_type: 'single', 'two-layer', or 'neural-net'

    Returns:
        Trained model coefficients
    """
    if len(samples) == 0:
        raise ValueError("No samples to train on")

    # Filter samples with concentration data
    samples_with_conc = [
        s for s in samples
        if len(s.concentrations) > 0 and any(c > 0 for c in s.concentrations.values())
    ]

    if len(samples_with_conc) == 0:
        raise ValueError("No samples with concentration data found")

    print(f"Filtered to {len(samples_with_conc)} samples with concentration data "
          f"(from {len(samples)} total)")

    # Detect fluorescent samples
    fluorescent_samples = [s for s in samples_with_conc if has_fluorescence(s.spectrum)]
    print(f"Found {len(fluorescent_samples)} fluorescent samples (R > 1.0)")

    # Get unique reagents
    reagent_set = set()
    for s in samples_with_conc:
        for reagent, conc in s.concentrations.items():
            if conc > 0:
                reagent_set.add(reagent)
    reagents = sorted(list(reagent_set))

    print(f"Training with {len(samples_with_conc)} samples and {len(reagents)} reagents: "
          f"{', '.join(reagents)}")

    # Build concentration matrix [Samples x Reagents]
    C = np.array([
        [s.concentrations.get(r, 0) / 100.0 for r in reagents]
        for s in samples_with_conc
    ])

    if model_type == 'single':
        return train_single_layer_model(samples_with_conc, reagents, C,
                                        len(fluorescent_samples) > 0)
    elif model_type == 'two-layer':
        return train_two_layer_model(samples_with_conc, reagents, C,
                                      len(fluorescent_samples) > 0)
    elif model_type == 'neural-net':
        return train_neural_net_model(samples_with_conc, reagents, C)
    else:
        raise ValueError(f"Unknown model type: {model_type}")


def train_single_layer_model(samples: List[SampleData], reagents: List[str],
                             C: np.ndarray, has_fluorescent: bool) -> ModelCoefficients:
    """Train single-layer Kubelka-Munk model"""

    # Separate reflectance and fluorescence components
    reflectance_spectra = []
    fluorescence_spectra = []

    for s in samples:
        refl, fluor = extract_fluorescence(s.spectrum)
        reflectance_spectra.append(refl)
        fluorescence_spectra.append(fluor)

    reflectance_spectra = np.array(reflectance_spectra)
    fluorescence_spectra = np.array(fluorescence_spectra)

    # Convert reflectance to K/S
    KS_measured = reflectance_to_ks(reflectance_spectra)

    # Solve for alpha coefficients using pseudo-inverse
    try:
        C_pinv = pseudo_inverse(C)
        Alpha = C_pinv @ KS_measured  # [Reagents x Wavelengths]

        alpha_map = {reagent: Alpha[i, :] for i, reagent in enumerate(reagents)}

        # Train fluorescence model if needed
        fluor_map = None
        if has_fluorescent:
            print("Training fluorescence emission model...")
            Fluor = C_pinv @ fluorescence_spectra
            fluor_map = {reagent: Fluor[i, :] for i, reagent in enumerate(reagents)}
            print("Fluorescence model trained successfully")

        print("Single-layer model trained successfully")

        return ModelCoefficients(
            type='single',
            alpha=alpha_map,
            fluorescence=fluor_map,
            wavelengths=WAVELENGTHS
        )

    except Exception as e:
        print(f"Failed to compute pseudo-inverse: {e}")
        raise ValueError("Failed to train single-layer model. "
                        "Try using fewer samples or check data quality.")


def train_two_layer_model(samples: List[SampleData], reagents: List[str],
                          C: np.ndarray, has_fluorescent: bool) -> ModelCoefficients:
    """Train two-layer Kubelka-Munk model using gradient descent"""

    num_wavelengths = len(WAVELENGTHS)
    num_reagents = len(reagents)

    # Separate reflectance and fluorescence
    processed_samples = []
    for s in samples:
        refl, fluor = extract_fluorescence(s.spectrum)
        processed_samples.append({
            'reflectance': refl,
            'fluorescence': fluor,
            'thickness': s.thickness
        })

    K_map = {r: np.zeros(num_wavelengths) for r in reagents}
    S_map = {r: np.zeros(num_wavelengths) for r in reagents}
    F_map = {r: np.zeros(num_wavelengths) for r in reagents} if has_fluorescent else None

    # Optimize for each wavelength independently
    for w in range(num_wavelengths):
        R_targets = np.array([s['reflectance'][w] for s in processed_samples])
        X_vals = np.array([s['thickness'] for s in processed_samples])

        # Initial guess: K ~ 1.0, S ~ 10.0
        params = np.concatenate([
            np.ones(num_reagents),      # K values
            np.ones(num_reagents) * 10  # S values
        ])

        # Gradient descent with momentum
        initial_lr = 0.001
        iterations = 200
        momentum = 0.9
        velocity = np.zeros(2 * num_reagents)

        for iter in range(iterations):
            grads = np.zeros(2 * num_reagents)

            # Compute gradients
            for s_idx in range(len(processed_samples)):
                concs = C[s_idx]
                X = X_vals[s_idx]
                R_tgt = R_targets[s_idx]

                # Current K_mix, S_mix
                K_mix = np.dot(concs, params[:num_reagents])
                S_mix = np.dot(concs, params[num_reagents:])

                R_pred = two_layer_forward(K_mix, S_mix, Rg_DEFAULT, X)
                error = R_pred - R_tgt

                # Finite difference approximation
                epsilon = 1e-4

                for r in range(num_reagents):
                    # dL/dK_r
                    R_k_plus = two_layer_forward(K_mix + epsilon * concs[r], S_mix, Rg_DEFAULT, X)
                    dR_dK = (R_k_plus - R_pred) / epsilon
                    grads[r] += 2 * error * dR_dK

                    # dL/dS_r
                    R_s_plus = two_layer_forward(K_mix, S_mix + epsilon * concs[r], Rg_DEFAULT, X)
                    dR_dS = (R_s_plus - R_pred) / epsilon
                    grads[r + num_reagents] += 2 * error * dR_dS

            # Normalize gradients
            grad_norm = np.linalg.norm(grads)
            if grad_norm > 1:
                grads = grads / grad_norm

            # Update with momentum
            lr = initial_lr / (1 + iter / 100)
            velocity = momentum * velocity - lr * grads
            params += velocity

            # Enforce bounds
            params[:num_reagents] = np.maximum(0.01, params[:num_reagents])  # K >= 0.01
            params[num_reagents:] = np.maximum(0.1, params[num_reagents:])   # S >= 0.1

            # Early stopping
            if grad_norm < 1e-6 and iter > 50:
                break

        # Store optimized values
        for r_idx, reagent in enumerate(reagents):
            K_map[reagent][w] = params[r_idx]
            S_map[reagent][w] = params[r_idx + num_reagents]

        # Log progress
        if w % 10 == 0:
            avg_K = np.mean(params[:num_reagents])
            avg_S = np.mean(params[num_reagents:])
            print(f"λ={WAVELENGTHS[w]}nm: avg K={avg_K:.3f}, avg S={avg_S:.3f}")

        # Train fluorescence separately
        if has_fluorescent:
            F_targets = np.array([s['fluorescence'][w] for s in processed_samples])
            C_pinv = pseudo_inverse(C)
            F_coeffs = (C_pinv @ F_targets.reshape(-1, 1)).flatten()
            for r_idx, reagent in enumerate(reagents):
                F_map[reagent][w] = max(0, F_coeffs[r_idx])

    print("Two-layer model trained successfully")

    return ModelCoefficients(
        type='two-layer',
        K=K_map,
        S=S_map,
        fluorescence=F_map,
        wavelengths=WAVELENGTHS
    )


def train_neural_net_model(samples: List[SampleData], reagents: List[str],
                           C: np.ndarray, use_mixup: bool = True,
                           use_kfold: bool = False) -> ModelCoefficients:
    """Train neural network model with optional augmentation"""

    print("Training neural network (black-box model)...")

    # Apply mix-up augmentation if requested
    if use_mixup and len(samples) >= 2:
        from utils.data_augmentation import create_mixup_augmented_dataset
        augmented_samples = create_mixup_augmented_dataset(
            samples,
            n_synthetic=min(20, len(samples) * 4),  # 4x augmentation
            lambda_min=0.2,
            lambda_max=0.8
        )

        # Rebuild concentration matrix for augmented data
        C_aug = np.array([
            [s.concentrations.get(r, 0) / 100.0 for r in reagents]
            for s in augmented_samples
        ])
        X = np.column_stack([C_aug, np.array([s.thickness for s in augmented_samples])])
        Y = np.array([s.spectrum for s in augmented_samples])
    else:
        X = np.column_stack([C, np.array([s.thickness for s in samples])])
        Y = np.array([s.spectrum for s in samples])

    # Adjust dropout based on dataset size
    dropout_rate = 0.1 if len(samples) < 20 else 0.2

    # Train neural network with improvements (4×32 deep architecture)
    nn_data = train_neural_network(
        X, Y,
        hidden_size=32,
        learning_rate=0.005,
        epochs=2000,
        batch_size=min(8, max(4, len(samples) // 5)),
        l2_lambda=0.005,
        dropout_rate=dropout_rate,
        use_kfold=use_kfold
    )

    return ModelCoefficients(
        type='neural-net',
        neural_net={
            'reagents': reagents,
            'model': nn_data['model'],
            'input_mean': nn_data['input_mean'],
            'input_std': nn_data['input_std'],
            'output_mean': nn_data['output_mean'],
            'output_std': nn_data['output_std']
        },
        wavelengths=WAVELENGTHS
    )


# --- Prediction Functions ---

def predict_reflectance(concentrations: Dict[str, float],
                       model: ModelCoefficients,
                       thickness: float = 4.0) -> np.ndarray:
    """
    Predict reflectance spectrum for given formulation

    Args:
        concentrations: Dictionary of reagent concentrations (%)
        model: Trained model coefficients
        thickness: Film thickness

    Returns:
        Predicted reflectance spectrum
    """
    if model.type == 'neural-net' and model.neural_net:
        # Neural network prediction
        nn = model.neural_net
        reagents = nn['reagents']

        # Prepare input
        input_data = np.array([concentrations.get(r, 0) / 100.0 for r in reagents] + [thickness])

        return predict_neural_network(
            input_data,
            nn['model'],
            nn['input_mean'],
            nn['input_std'],
            nn['output_mean'],
            nn['output_std']
        )

    # K-M based predictions
    num_wavelengths = len(model.wavelengths)
    predicted_R = np.zeros(num_wavelengths)

    for i in range(num_wavelengths):
        if model.type == 'single':
            # Single Layer K-M
            ks_sum = 0.0
            for reagent, percent in concentrations.items():
                if model.alpha and reagent in model.alpha:
                    unit_ks = model.alpha[reagent][i]
                    ks_sum += (percent / 100.0) * unit_ks

            ks_sum = max(0.0001, ks_sum)
            base_reflectance = ks_to_reflectance(np.array([ks_sum]))[0]

        else:  # two-layer
            k_sum = 0.0
            s_sum = 0.0
            for reagent, percent in concentrations.items():
                if model.K and reagent in model.K:
                    unit_k = model.K[reagent][i]
                    k_sum += (percent / 100.0) * unit_k
                if model.S and reagent in model.S:
                    unit_s = model.S[reagent][i]
                    s_sum += (percent / 100.0) * unit_s

            base_reflectance = two_layer_forward(k_sum, s_sum, Rg_DEFAULT, thickness)

        # Add fluorescence contribution
        fluorescence = 0.0
        if model.fluorescence:
            for reagent, percent in concentrations.items():
                if reagent in model.fluorescence:
                    unit_f = model.fluorescence[reagent][i]
                    fluorescence += (percent / 100.0) * unit_f

        predicted_R[i] = base_reflectance + fluorescence

    return predicted_R
